src/Component.php

```diff
@@ -187,14 +201,19 @@ protected function getVariantDefaultVariables(string $modifier, array $defined_v
    * @return array
    *   An array with three elements:
    *   1. the component's pathname
-   *   2. the component basename without variants
-   *   3. a list of variants, if any.
+   *   2. the component's compound pathname
+   *   3. the component basename without variants
+   *   4. a list of variants, if any.
    */
   protected function extractParts(string $compound_name): array {
-    $pathname = preg_replace('@--[^.]+@', '', $compound_name);
+    $library = Drupal::service('twig.loader.componentlibrary');
     $variants = explode('--', basename(basename($compound_name, '.twig'), '.html'));
     $component = array_shift($variants);
-    return [$pathname, $component, $variants];
+    $pathname = $templatePathname = preg_replace('@--[^.]+@', '', $compound_name);
+    if ($library->exists($compound_name)) {
+      $templatePathname = $compound_name;
```

Comment:
This would be a bit more readable & clear:

```php
    $pathname = preg_replace('@--[^.]+@', '', $compound_name);
    $templatePathname = $library->exists($compound_name) ? $compound_name : $pathname;
```
I'd also move the two assignments back to the beginning of the method to match the order of return values and docs.